### ISSUE TITLE
Allow SQL cells to access exported python functions as UDFs

### ISSUE THREAD

**Issue by okennedy** [VizierDB/web-ui#225]
Recent versions of the Vizier-API allow python functions to be exported for import/use in later cells.  Currently this export only supports python cells.  It would be nifty if we could export python functions to be invoked from a SQL cell as a UDF.  Given that spark supports python function execution, this should be feasible, but some study will be required to figure out exactly how to do it.

**Comment by okennedy** [comment]
Privatization may be making this a little more difficult than it needs to be, so I'm going to start by cataloguing what I know.  

SparkSQL has a [PythonUDF expression](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala), which seems to take in some basic details: name, type details, arguments, etc...  The UDF references a [PythonFunction object](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala#L76) which, unfortunately, is private to the spark package.  

The PythonFunction object is defined as:
* **command**: A sequence of bytes encoding the function itself? (function body?  function bytecode?  serialized python?)
* **envVars**: A java map.  `sys.env`?
* **pythonIncludes**: A list of strings.  Probably just the included libraries
* **pythonExec**: ???  The python executable?  
* **pythonVer**: String.  3.0?
* **broadcastVars**: A java list.  Global variables?
* **accumulator**: One of [these jobbies](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala#L683).  Presumably for aggregates

Function registration happens through [UDFRegistration.registerPython](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala#L51) which is a shorthand for the normal UDF registration (`spark.udf.functionRegistry.createOrReplaceTempFunction(name, Seq[Expression] => Expression`) and [UserDefinedPythonFunction](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala), which is a utility class that creates `PythonUDF`s

Assuming we can reverse engineer PythonFunction, this should be relatively straightforward.

**Comment by okennedy** [comment]
From the Vizier side, `web-api-async` uses Mimir's Blob storage to store Python UDFs.  Here's an example of what a function would look like:
```
@vizierdb.export_module_decorator
def apply_foo(a):
    return a + 1
```

**Comment by okennedy** [comment]
A few more notes.

Python execution seems to start with [PythonRunner](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala#L78), which passes `pythonExec` (the path to the python binary), to [PythonWorkerFactory](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala).  This, in turn, starts up the python worker module (by default: [pyspark.worker](https://github.com/apache/spark/blob/master/python/pyspark/worker.py)) which connects via a py4j gateway.

In the worker, the actual execution happens [here](https://github.com/apache/spark/blob/d378dc5f6db6fe37426728bea714f44b94a94861/python/pyspark/worker.py#L581), where the function reads a UDF through pickle (actually seems like it reads a pickled instance of the [UDF wrapper class](https://github.com/apache/spark/blob/master/python/pyspark/sql/udf.py)

More precisely, it looks like python UDFs are getting serialized [here](https://github.com/apache/spark/blob/d378dc5f6db6fe37426728bea714f44b94a94861/python/pyspark/rdd.py#L2515), using [CloudPickleSerializer](https://github.com/apache/spark/blob/d378dc5f6db6fe37426728bea714f44b94a94861/python/pyspark/serializers.py#L430), which is the head of a long chain of dependencies to [CloudPickler](https://github.com/apache/spark/blob/d378dc5f6db6fe37426728bea714f44b94a94861/python/pyspark/cloudpickle/cloudpickle_fast.py#L458).  

In short, what we need to get a PythonFunction is call [cloudpickle.dumps()](https://github.com/apache/spark/blob/d378dc5f6db6fe37426728bea714f44b94a94861/python/pyspark/cloudpickle/__init__.py) on the function.

Actual execution happen

**Comment by okennedy** [comment]
Since Vizier dumps out Python code in plaintext, we'd need to go through python anyway to pickle the result.  Perhaps we don't actually want to do this... 

One approach might be to just run cloudpickle.dumps() once to get a "magic" serialized function:
```

def invoke_vizier_udf(*args):
  fn = args[0]
  args = args[1:]
  class VizierUDFWrapper:
    def export_module_decorator(fn):
      self.fn = fn
      return fn
  vizierdb = VizierUDFWrapper()
  exec(fn)
  return vizierdb.fn(*args)
```

**Comment by okennedy** [comment]
![image](https://user-images.githubusercontent.com/3641553/96525646-8482a880-1249-11eb-8a36-efdf514e0059.png)
